### PR TITLE
Reverse DNS records for named headless services

### PR DESCRIPTION
**Pull request by sadlil**
Add reverse DNS records for pods' IPs to named headless services' with Pod Hosts FQDN `(<podHostName>.<serviceName>.<namespace>.svc.cluster.local)`.

Resolves issue #33470

Implemented this by watching endpoints, If the service is used in endpoints Referenced Pods host name a reverse DNS entry is created. Referenced Pods is getting via an InCluster kube API call.

### CODE REVIEW

**Comment by sadlil** [comment]
I signed it!

**Comment by sadlil** [comment]
cc @harryge00 @smarterclayton

**Comment by smarterclayton** [comment]
@thockin in think this is the change where we create a doc with the DNS schema and start keeping it up to date.  Kube-DNS code is unreadable for this and I think we owe it to have a real spec.

I'm provisionally ok with this since we've tacit agreement that reverse entries to services and endpoint are ok, and a reasonable service client would want this info, and it can also be implemented in O(1) for services by doing an endpoint fetch.

**Comment by smarterclayton** [comment]
@sadlil can you add details to your description formally defining what queries you will answer with more examples?  That'll help when we define schema and review tests

**Comment by smarterclayton** [comment]
@k8s-bot ok to test

**Comment by sadlil** [comment]
@smarterclayton, there is already a PTR record entry for the service ip to service domain. This PR adds PTR entry for pods IP for only in case of named headless service. With a named headless, A records resolved for Pod IP via `podHostName.headlessServiceName.namespace.svc.cluster.local`, But the PTR records for those Pods' IP didn't resolve to that FQDN.
This PR adds PTR records for Pod's IPs to the FQDN, only the Pod have a **Named Headless Service**.
This PR should response for Reverse record for pods with its FQDN.

**Comment by smarterclayton** [comment]
Because this is an API change, I'm asking for a more rigorous description of the change as the description of this PR so that we can review the design and the implementation.  Try to expand the description to include an example and a bit more information about the exact queries a DNS consumer would make.

**Comment by thockin** [comment]
@smarterclayton you mean this spec: https://github.com/kubernetes/kubernetes/pull/38408 ?  It's a little aspirational in that it includes a few things we don't implement - such as this change :)

PTAL at spec proposal, I'll look at this PR

File: pkg/dns/dns.go [hunk 1]
```diff
@@ -330,6 +328,32 @@ func (kd *KubeDNS) handleEndpointAdd(obj interface{}) {
 	}
 }
 
+func (kd *KubeDNS) handleEndpointUpdate(oldObj, newObj interface{}) {
+	kd.handleEndpointDelete(oldObj)
```
Comment (thockin) [inline pkg/dns/dns.go]: This gives us a brief window during which resolution will fail.  We can't do that.  This seems to be a logically distinct change, please do a different PR if you want to tackle this.

File: pkg/dns/dns.go [hunk 2]
```diff
@@ -433,6 +464,28 @@ func (kd *KubeDNS) generateRecordsForHeadlessService(e *v1.Endpoints, svc *v1.Se
 	return nil
 }
 
+func (kd *KubeDNS) isHeadlessServiceNamed(svc *v1.Service, podReference *v1.ObjectReference) bool {
+	if podReference != nil {
+		pod, err := kd.kubeClient.Core().Pods(podReference.Namespace).Get(podReference.Name)
```
Comment (thockin) [inline pkg/dns/dns.go]: We should not be accessing API for pods in DNS.  We Should have enough information in the Endpoints structure - if the EndpointAddress has a .Hostname set *and* the Service is headless, we should just take it, I think.  It's a little sloppy, but I am worried about the churn of doing this for every backend of every headless service.

Maybe we can pre-filter by the existence of EndpointAddress.Hostname, so we only need to do this on those endpoints that elect themselves to have a hostname.  Still seems like a DoS vector.  @smarterclayton  @bowei

**Comment by sadlil** [comment]
@thockin i have removed the api access for finding the named headless services from the end point. As you said and [endpoint_controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L425) shows that an EndpointAddress.Hostname is set then the service is named in Pod.
I am still confused about `kd.handleEndpointDelete(oldObj)`.

**Comment by sadlil** [comment]
@thockin [Commit f6c3b8](https://github.com/kubernetes/kubernetes/pull/38376/commits/f6c3b8dcb4ec546645ac0b5a12e0dd18487bc3b5) should reduce the "window during which resolution will fail". But still some unwanted updates are there.

@smarterclayton 
Type: PTR, Name: <reverse-ip-name>

If a ready endpoint exists with IP equal to the IP of <reverse-ip-name> exists, and the endpoint has a non-empty hostname with value <hostname>, then a PTR record with <hostname>.<service>.<ns>.svc.<zone> will be returned.

**Comment by smarterclayton** [comment]
@thockin I want to start (with this PR) ensuring every API change to DNS is
recorded in an API design doc of some form.  Review up front is easier,
review in the future is easier (this doesn't fit the spec), and downstream
implementations like OpenShift and the proposed CoreDNS is easier.

**Comment by sadlil** [comment]
Ping @thockin

**Comment by bowei** [comment]
@sadlil can you address the build failures or diagnose them

**Comment by sadlil** [comment]
@bowei GKE smoke e2e is failing now only. Looks like its not required. If any more changes needed let me know.

**Comment by TamalSaha** [comment]
@k8s-bot cvm gke e2e test this

**Comment by sadlil** [comment]
@k8s-bot cvm gke e2e test this

**Comment by sadlil** [comment]
@bowei rebase fixed the build failures.

**Comment by thockin** [comment]
The spec that @smarterclayton wants (I think) is #38408 and is not yet merged.  Can you write this up as if it were a patch on that doc?

**Comment by xialonglee** [comment]
I just passed by and curious about one thing that:
if i create a headless service `serviceA` and `serviceB` without selector,
then i create endpoint `endpointA` and `endpointB` respectively.
the point is that if both  `endpointA` and  `endpointB` share the same subset content , e.g. 
`    "subsets": [
        {
            "addresses": [
                { "ip": "1.2.3.4" }
            ],
            "ports": [
                { "port": 9376 }
            ]
        }
    ]
`
i tested , it could be done and kube-dns can resolve these A-records of both `serviceA.namespace.svc.cluster.local` and `serviceB.namespace.svc.cluster.local` and return IP `1.2.3.4`
so if i searched `nslookup 1.2.3.4` , which PTR record should return?

**Comment by TamalSaha** [comment]
@xialonglee, this sets PTR record for "named" headless service. If you check http://kubernetes.io/docs/admin/dns/, the pods backing the endpoint needs to set `subdomain` that matches a headless service. So, there can be only one named headless service for a pod.

**Comment by xialonglee** [comment]
@TamalSaha Thanks for answering me. 
So a pod with `subdomain` field only matches the service that named as same as subdomain.
I am all cleared now :)

**Comment by TamalSaha** [comment]
@xialonglee you got it.

**Comment by harryge00** [comment]
I think this PR also solves https://github.com/kubernetes/kubernetes/issues/26752?

**Comment by TamalSaha** [comment]
@sadlil , you need to reopen Pr against https://github.com/kubernetes/dns/blob/master/pkg/dns/dns.go :(

**Comment by sadlil** [comment]
This PR is opened in kubernetes/dns#25. Should this be closed?
